pkg/controller/autoscaling/elasticsearch/autoscaler/horizontal.go

```diff
@@ -0,0 +1,115 @@
+
+package autoscaler
+
+import (
+	"fmt"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/autoscaling/elasticsearch/resources"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/autoscaling/elasticsearch/status"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// scaleHorizontally adds or removes nodes in a set of node sets to provide the required capacity in a tier.
+func (ctx *Context) scaleHorizontally(
+	nodeCapacity resources.NodeResources, // resources for each node in the tier/policy, as computed by the vertical autoscaler.
+) resources.NodeSetsResources {
+	minNodes := int(ctx.AutoscalingSpec.NodeCount.Min)
+	maxNodes := int(ctx.AutoscalingSpec.NodeCount.Max)
+	totalRequiredCapacity := ctx.RequiredCapacity.Total // total required resources, at the tier level.
+	nodeToAdd := 0
+
+	// Scale horizontally to match memory requirements
+	if !totalRequiredCapacity.Memory.IsZero() {
+		nodeMemory := nodeCapacity.GetRequest(corev1.ResourceMemory)
+		nodeToAdd = ctx.getNodesToAdd(nodeMemory.Value(), totalRequiredCapacity.Memory.Value(), minNodes, maxNodes, string(corev1.ResourceMemory))
+	}
+
+	// Scale horizontally to match storage requirements
+	if !totalRequiredCapacity.Storage.IsZero() {
+		nodeStorage := nodeCapacity.GetRequest(corev1.ResourceStorage)
+		nodeToAdd = max(nodeToAdd, ctx.getNodesToAdd(nodeStorage.Value(), totalRequiredCapacity.Storage.Value(), minNodes, maxNodes, string(corev1.ResourceStorage)))
+	}
+
+	totalNodes := nodeToAdd + minNodes
+	ctx.Log.Info("Horizontal autoscaler", "policy", ctx.AutoscalingSpec.Name,
+		"scope", "tier",
+		"count", totalNodes,
+		"required_capacity", totalRequiredCapacity,
+	)
+
+	nodeSetsResources := resources.NewNodeSetsResources(ctx.AutoscalingSpec.Name, ctx.NodeSets.Names())
+	nodeSetsResources.NodeResources = nodeCapacity
+	fnm := NewFairNodesManager(ctx.Log, nodeSetsResources.NodeSetNodeCount)
+	for totalNodes > 0 {
+		fnm.AddNode()
+		totalNodes--
+	}
+
+	return nodeSetsResources
+}
+
+// getNodesToAdd calculates the number of nodes to add in order to comply with the capacity requested by Elasticsearch.
+func (ctx *Context) getNodesToAdd(
+	nodeResourceCapacity int64, // resource capacity of a single node, for example the memory of a node in the tier
+	totalRequiredCapacity int64, // required capacity at the tier level
+	minNodes, maxNodes int, // min and max number of nodes in this tier, as specified by the user the autoscaling spec.
+	resourceName string, // used for logging and in events
+) int {
+	// minResourceQuantity is the resource quantity in the tier before scaling horizontally.
+	minResourceQuantity := int64(minNodes) * nodeResourceCapacity
+	// resourceDelta holds the resource needed to comply with what is requested by Elasticsearch.
+	resourceDelta := totalRequiredCapacity - minResourceQuantity
+	// getNodeDelta translates resourceDelta into a number of nodes.
+	nodeToAdd := getNodeDelta(resourceDelta, nodeResourceCapacity)
```

Comment:
This is actually the number of nodes that must be added to fulfill the requirements, I tried to improve the comment [here](https://github.com/elastic/cloud-on-k8s/pull/4173/files#diff-e0e0ede710eb10e3e4b3e10b8aa6f25802dc7a18a4784fa66b16410613de6b4dR21).

For example, if there is a minimum number of 2 (because this is what the user specified in the spec) but we need a total of 5 nodes in the cluster then `nodeToAdd` is 3